### DISCUSSION TITLE
[General]Unable to use Paymaster on iOS with WalletConnect

### DISCUSSION THREAD

**Discussion by equalogix** [zkSync-Community-Hub/zksync-developers#53] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

Before starting, I am @BlindChicken on discord.
I'm trying to use a paymaster to pay for transactions in my dApp. The dApp must work on mobile and thus requires WalletConnect integration.

Note: the problem below occurs on both Testnet and Mainnet. (you should probably add a option "both")

### What is the specific issue or error you're encountering?

The issue is that it is not possible to send transactions that use a paymaster when on iOS and using WalletConnect. This is a bit of an edge case, but I believe it is of critical importance, because I don't see how one can offer a consumer application that does not work on iPhones.

Let me add some background, I have been struggling with this issue for a while, but I had found a bit of hacked solution to be able to pass transactions that use a paymaster using WalletConnect/Wagmi.sh. Unfortunately, last week WalletConnect sunsetted its V1, and my hack relied on V1.

As  result I decided to write my own version of "Wagmi" (React Context and custom hooks), such that it would be compatible with zkSync/Paymaster transactions and would continue to use Ethers.js. As such I have been doing a lot more thorough testing on a wider range of devices. The new code worked, fine on all devices except for on iOS.

After some research I found a documented issue with WalletConnect `iOS App Link Constraints` see https://docs.walletconnect.com/2.0/ios/guides/mobile-linking#ios-app-link-constraints for details. There is a proposed work around that relies on [signer.sendUncheckedTransaction](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/#JsonRpcSigner-sendUncheckedTransaction). This work-around does solve the problem of the `DeepLinks` in iOS but when a transaction is sent like this it ignores the paymaster and asks the user to include the fees in the transaction.

This is expected based on the ethers.js docs, here is an excerpt:
`The JSON-RPC API only provides a transaction hash as the response when a transaction is sent, but the ethers Provider requires populating all details of a transaction before returning it. For example, the gas price and gas limit may be adjusted by the node or the nonce automatically included, in which case the opaque transaction hash has discarded this.`
My understanding (I may be wrong?) is that it is exactly at this step where the gas price would be adjusted to account for the use of the paymaster. Using `sendUncheckedTransaction` skips this step and thus skips the paymaster, and as a result one cannot use a paymaster on iOS with WalletConnect.

I believe that this is critical bug, because if one cannot use a paymaster when on iOS/iPhone then its value is very much limited (read 0). And, I can't expect my users "normies" to start opening and trying to use my dApp in the MetaMask browser, it's bad enough they need to download the wallet and set it up.

That being said given the nature of this bug and that it occurs at the intersection of four different technologies, I'm not sure if this is an issue with zkSync, WalletConnect or ethers.js. But the impact definitely on zkSync.

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

NA

### Are there any external libraries or contracts that your contract interacts with?

```
import { Web3Provider, Provider, Contract } from "zksync-web3";         // version "^0.14.3"
import { EthereumProvider } from '@walletconnect/ethereum-provider' // version "^2.8.6"
import { WalletConnectModal } from '@walletconnect/modal'                // version "^2.5.9"
import { ethers } from "ethers";                                                               // version "^5.7.2"
```

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

see above

### What steps have you already taken to try to resolve the issue?

see above

**Comment by equalogix** [comment]
Any Updates on this?

**Reply by MexicanAce** [reply, marked as answer]
Latest responses can be found in https://github.com/zkSync-Community-Hub/zkync-developers/discussions/63

**Comment by caydiverlarachel** [comment]
This issue indeed seems to be at the intersection of multiple technologies, including zkSync, WalletConnect, and ethers.js. To address this critical bug, thorough testing and coordination between the respective development teams may be necessary. It's essential to ensure compatibility and functionality across different platforms, especially for widely used devices like iPhones. Consider reaching out to the support or development teams of these technologies for further assistance and resolution of the issue.

**Comment by pobertester** [comment]
Environment: Testnet
Issue: Cannot use paymaster for transactions on iOS/iPhone when using WalletConnect
Impact: Critical, as it limits the usability of dApps on iOS/iPhone devices
Affected Technologies: zkSync, WalletConnect, ethers.js
This issue highlights the critical importance of enabling paymaster functionality for transactions on iOS/iPhone devices when using WalletConnect, as it significantly impacts the usability and adoption of dApps on these platforms.

**Comment by dariuscandfield** [comment]
This issue seems to be at the intersection of zkSync, WalletConnect, ethers.js, and possibly iOS limitations. It appears that using sendUncheckedTransaction bypasses the paymaster, leading to a lack of fee coverage on iOS devices. Resolving this may require collaboration between the involved technologies to find a suitable workaround or solution.
